File: 05-Lista_simplemente_ligada/cpp/simpleLinkedList.cpp
#include "SimpleLinkedList.h"

template <typename T>
SimpleLinkedList<T>::~SimpleLinkedList() {
    delete head;
}

template <typename T>
bool SimpleLinkedList<T>::isEmpty(){
    return head == nullptr;
}

template <typename T>
int SimpleLinkedList<T>::getSize(){
    int size = 0;
    Node<T> *iteratorNode = head;
    while(iteratorNode != nullptr){
        size++;
        iteratorNode = iteratorNode->getNext();
    }
    return size;
}

template <typename T>
int SimpleLinkedList<T>::search(T valor){
    int index = 0;
    Node<T> *iteratorNode = head;
    while(iteratorNode != nullptr){
        if(iteratorNode->getData() == valor){
            return index;
        }
        index++;
        iteratorNode = iteratorNode->getNext();
    }
    return -1;
}

template <typename T>
void SimpleLinkedList<T>::addLast(T valor){
    Node<T> *newNode = new Node<T>(valor);
    if(isEmpty()){
        head = newNode;
        return;
    }

    Node<T> *iteratorNode = head;
    while(iteratorNode->getNext() != nullptr){
        iteratorNode = iteratorNode->getNext();
    }
    iteratorNode->setNext(newNode);
}

template <typename T>
void SimpleLinkedList<T>::addFirst(T valor){
    Node<T> *newNode = new Node<T>(valor, head);
    head = newNode;
}

template <typename T>
void SimpleLinkedList<T>::addAfter(T referencia, T valor){
    if(search(referencia) < 0) return;

    Node<T> *iteratorNode = head;
    while(iteratorNode->getNext() != nullptr && iteratorNode->getData() != referencia){
        iteratorNode = iteratorNode->getNext();
    }
    Node<T> *newNode = new Node(valor, iteratorNode->getNext());
    iteratorNode->setNext(newNode);
}

template <typename T>
void SimpleLinkedList<T>::deleteAt(int index){
    if(index < 0 || index >= getSize() || isEmpty()) return;

    if(index == 0){
        deleteStart();
        return;
    }

    int indexPosition = 0;
    Node<T> *iteratorNode = head;
    Node<T> *prevNode = nullptr;
    while(iteratorNode->getNext() != nullptr && indexPosition < index){
        prevNode = iteratorNode;
        iteratorNode = iteratorNode->getNext();
        indexPosition++;
    }
    prevNode->setNext(iteratorNode->getNext());
    iteratorNode->setNext(nullptr);
    delete iteratorNode;
}

template <typename T>
void SimpleLinkedList<T>::deleteStart(){
    if(isEmpty()) return;

    Node<T> *aux = head;
    head = head->getNext();
    aux->setNext(nullptr);
    delete aux;
}

template <typename T>
void SimpleLinkedList<T>::deleteLast(){
    if(isEmpty()) return;

    if(getSize() == 1){
        deleteStart();
        return;
    }

    Node<T> *iteratorNode = head;
    Node<T> *prevNode = nullptr;
    while(iteratorNode->getNext() != nullptr){
        prevNode = iteratorNode;
        iteratorNode = iteratorNode->getNext();
    }
    prevNode->setNext(nullptr);
    delete iteratorNode;
}

template <typename T>
void SimpleLinkedList<T>::update(T a_buscar, T valor){
    if(search(a_buscar) < 0) return;

    Node<T> *iteratorNode = head;
    while(iteratorNode->getNext() != nullptr && iteratorNode->getData() != a_buscar){
        iteratorNode = iteratorNode->getNext();
    }
    iteratorNode->setData(valor);
}

File: 05-Lista_simplemente_ligada/python/simpleLinkedList.py
from matplotlib.ticker import NullFormatter
from requests import head
from node import Node

class SimpleLinkedList:
    head = None

    def __init__(self):
        head = None

    def isEmpty(self):
        return self.head == None

    def getSize(self):
        current = self.head
        count = 0

        while current != None:
            count += 1
            current = current.getNext()
        return count

    def search(self, valor):
        indexPosition = 0
        current = self.head
        while current != None:
            if current.getData() == valor:
                return indexPosition
            current = current.getNext()
            indexPosition += 1
        return -1


    def addLast(self, valor):
        if self.isEmpty():
            self.head = Node(valor)
            return

        current = self.head
        while current.getNext() != None:
            current = current.getNext()
        temp = Node(valor)
        current.setNext(temp)

    def addFirst(self, valor):
        newNode = Node(valor, self.head)
        self.head = newNode

    def addAfter(self, referencia, valor):
        if(self.search(referencia) < 0):
            return

        current = self.head
        while current != None and current.getData() != referencia:
            current = current.getNext()
        newNode = Node(valor, current.getNext())
        current.setNext(newNode)

    def deleteAt(self, index):
        if index < 0 or index >= self.getSize() or self.isEmpty():
            return
        if index == 0:
            self.deleteStart()
            return

        indexPosition = 0
        prevNode = None
        current = self.head
        while current != None and indexPosition != index:
            prevNode = current
            current = current.getNext()
            indexPosition += 1
        prevNode.setNext(current.getNext())

    def deleteStart(self):
        if self.isEmpty():
            return
        self.head = self.head.getNext()

    def deleteLast(self):
        if self.isEmpty():
            return
        if self.getSize() == 1:
            self.deleteStart()
            return

        prev = None
        current = self.head
        while current.getNext() != None:
            prev = current
            current = current.getNext()
        prev.setNext(None)

    def update(self, a_buscar, valor):
        if self.search(a_buscar) < 0:
            return

        current = self.head
        while current != None and current.getData() != a_buscar:
            current = current.getNext()
        current.setData(valor)

    def __str__(self):
        current = self.head
        cadena = ""
        while current != None:
            cadena += str(current)
            current = current.getNext()
        return cadena
